File: 206-reverse-linked-list.cpp
/*
 Source : https://leetcode.com/problems/reverse-linked-list/
 Author : Alex Xu
 Date   : May 28, 2015
 Problem:
 Reverse a singly linked list.

 Hint:
 A linked list can be reversed either iteratively or recursively. Could you implement both?
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == nullptr)
            return nullptr;
            
        ListNode* pre = nullptr;
        ListNode* next = nullptr;
        do {
            next = head->next;
            head->next = pre;
            pre = head;
            head = next;
        } while (head != nullptr);
        
        return pre;
    }
};

File: java/206-reverse-linked-list.java
/*
  Source : https://oj.leetcode.com/problems/reverse-linked-list
  Author : Alex Xu
  Date   : Jun 12, 2018
  Problem: Reverse a singly linked list.

Example:

Input: 1->2->3->4->5->NULL
Output: 5->4->3->2->1->NULL
Follow up:

A linked list can be reversed either iteratively or recursively. Could you implement both?
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

// Iterative solution: Runtime 0ms, beats 100%
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode dummy = new ListNode(0);
        ListNode next = null;
        while(head != null) {
            next = head.next;
            head.next = dummy.next;
            dummy.next = head;
            head = next;
        }
        
        return dummy.next;
    }
}

// Recursive solution: Runtime 2ms, beats 3.81%
class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        
        ListNode newHead = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        
        return newHead;
    }
}
